### PR TITLE
Fix codeowner detected diff base branch to mergebase

### LINKED ISSUE #29763: CODEOWNERS test against diff towards HEAD of target branch instead of merge-base
### Description

When a new PR is opened the CODEOWNERS functionality tests against the diff towards refs/heads/<target branch> as seen here:
https://github.com/go-gitea/gitea/blob/7fd0a5b276aadcf88dcc012fcd364fe160a58810/models/issues/pull.go#L920

This means that if the source branch for the PR is not up to date with the target the diff will contain all the changes added to the target branch since merge base - this seems like the wrong behavior to me.

I think it should use merge base instead of head - which i also assume is the logic used to determine which changes needs to be reviewed in a PR.

I am not familiar with the codebase of gitea, so i am not comfortable make the pr to change this, but if it is agreed that this change should be made, and nobody else has the bandwidth to do it, i might give it a try.

### Gitea Version

1.21.8

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Not relevant

### Database

MySQL/MariaDB

### PR DESCRIPTION
**Pull request by lunny**
Fix #29763

This PR fix 2 problems about CodeOwner in pull request.
- Don't use pull request base branch but merge base as a diff base to detect the codeowner.
- CodeOwner detection in fork repositories will be disabled because almost all the fork repositories will not change CODEOWNERS files but it should not be used on fork repositories' pull requests.

### CODE REVIEW

File: services/issue/pull.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	user_model "code.gitea.io/gitea/models/user"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/gitrepo"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+)
+
+func getMergeBase(repo *git.Repository, baseBranch, headBranch string) (string, error) {
+	// Add a temporary remote
+	tmpRemote := strconv.FormatInt(time.Now().UnixNano(), 10)
```
Comment (wxiaoguang) [inline services/issue/pull.go]: It's better to have a proper prefix for the temp remote name

File: services/issue/pull.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+	if err := repo.AddRemote(tmpRemote, repo.Path, false); err != nil {
+		return "", fmt.Errorf("AddRemote: %w", err)
+	}
+	defer func() {
+		if err := repo.RemoveRemote(tmpRemote); err != nil {
+			log.Error("GetPullRequestInfo: RemoveRemote: %v", err)
```
Comment (wxiaoguang) [inline services/issue/pull.go]: The error message doesn't match the function name

**Review by wxiaoguang — APPROVED** [review]
Overall LGTM, approve ahead

**Review by wolfogre — APPROVED** [review]

File: services/issue/pull.go (hunk 2 above)
Comment (lunny) [reply services/issue/pull.go]: 5425817

File: services/issue/pull.go (hunk 1 above)
Comment (lunny) [reply services/issue/pull.go]: d6b195eb45ec1e299c29c5da78af869d7c33f605

**Comment by silverwind** [comment]
Test failure related:

```
gitea/tests/integration/pull_review_test.go:36
        	Error:      	Received unexpected error:
        	            	initRepository: prepareRepoCommit: GetRepoInitFile[DEFAULT]: file does not exist
        	Test:       	TestPullView_CodeOwner
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x44b121b]

**Comment by lunny** [comment]
> Test failure related:
> 
> 
> 
> ```
> 
> gitea/tests/integration/pull_review_test.go:36
> 
>         	Error:      	Received unexpected error:
> 
>         	            	initRepository: prepareRepoCommit: GetRepoInitFile[DEFAULT]: file does not exist
> 
>         	Test:       	TestPullView_CodeOwner
> 
> panic: runtime error: invalid memory address or nil pointer dereference [recovered]
> 
> 	panic: runtime error: invalid memory address or nil pointer dereference
> 
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x44b121b]

It's very wired. It run successfully local.

File: tests/integration/pull_review_test.go [hunk 3]
```diff
@@ -20,3 +31,88 @@ func TestPullView_ReviewerMissed(t *testing.T) {
...
 }
+
+func TestPullView_CodeOwner(t *testing.T) {
+	onGiteaRun(t, func(t *testing.T, u *url.URL) {
+		user2 := unittest.AssertExistsAndLoadBean(t, &user_model.User{ID: 2})
+
+		// Create the repo.
+		repo, err := repo_service.CreateRepositoryDirectly(db.DefaultContext, user2, user2, repo_service.CreateRepoOptions{
+			Name:             "test_codeowner",
+			Readme:           "DEFAULT",
```
Comment (wxiaoguang) [inline tests/integration/pull_review_test.go]:
```suggestion
			Readme:           "Default",
```

**Pull request merged**